apps/aecore/src/aec_target.erl

```diff
@@ -28,40 +28,77 @@ determine_delta_header_height(Header) ->
-recalculate(Header, InitialHeader) ->
-    HeaderTarget = aec_headers:target(Header),
-    ExpectedRate = aec_governance:expected_block_mine_rate(),
-    CurrentRate = mining_rate_between(InitialHeader, Header),
-    aec_pow:recalculate_target(HeaderTarget, ExpectedRate, CurrentRate).
+-spec recalculate(header(), nonempty_list(header())) -> non_neg_integer().
+%% Target recalculation.
+%%
+%% Some concepts:
+%%
+%%    Difficulty = HIGHEST_TARGET / Target
+%%    Rate       = Capacity / Difficulty  (blocks/ms)
+%%    Capacity   = number of potential solutions per ms generated by miners
+%%
+%%    DesiredTimeBetweenBlocks = aec_governance:expected_block_mine_rate()
+%%    DesiredRate              = 1 / DesiredTimeBetweenBlocks
+%%
+%% The basic idea of the algorithm is to estimate the current network capacity
+%% based on the `N` (= 10) previous blocks and use that to set the new
+%% target:
+%%
+%%    NewDifficulty = EstimatedCapacity / DesiredRate
+%%    NewTarget     = HIGHEST_TARGET / NewDifficulty
+%%                  = HIGHEST_TARGET * DesiredRate / EstimatedCapacity
+%%
+%% We can estimate the network capacity used to mine a given block `i` as
+%%
+%%    EstimatedCapacity[i] = Difficulty[i] / MiningTime[i]
+%%    MiningTime[i]        = Time[i + 1] - Time[i]
+%%
+%% The estimated capacity across all `N` blocks is then the weighted (by time)
+%% average of the estimated capacities for each block.
+%%
+%%    EstimatedCapacity = Sum(EstimatedCapacity[i] * MiningTime[i]) / TotalTime
+%%                      = Sum(Difficulty[i]) / TotalTime
+%%                      = Sum(HIGHEST_TARGET / Target[i]) / TotalTime
+%%
+%% Now, the problem is that we can't do any floating point arithmetic (to
+%% ensure the calculation can be verified by other nodes), so we pick a
+%% reasonably big integer K (= HIGHEST_TARGET * 2^32) and compute
+%%
+%%    EstimatedCapacity ≈ Sum(K * HIGHEST_TARGET div Target[i]) / TotalTime / K
+%%
+%% Then
+%%
+%%    NewTarget = HIGHEST_TARGET * DesiredRate / EstimatedCapacity
+%%              ≈ HIGHEST_TARGET * DesiredRate * TotalTime * K / Sum(K * HIGHEST_TARGET div Target[i])
+%%              ≈ DesiredRate * TotalTime * K / Sum(K div Target[i])
+%%              ≈ TotalTime * K div (DesiredTimeBetweenBlocks * Sum(K div Target[i]))
+%%
+recalculate(Top, PrevHeaders) ->
+    K                        = aec_pow:scientific_to_integer(?HIGHEST_TARGET_SCI) * (1 bsl 32),
+    SumKDivTargets           = lists:sum([ K div aec_pow:scientific_to_integer(aec_headers:target(Hd))
+                                           || Hd <- PrevHeaders ]),
+    DesiredTimeBetweenBlocks = aec_governance:expected_block_mine_rate(),
+    Last                     = lists:last(PrevHeaders),
+    TotalTime                = mining_time_between(Last, Top),
+    NewTargetInt             = TotalTime * K div (DesiredTimeBetweenBlocks * SumKDivTargets),
+    min(?HIGHEST_TARGET_SCI, aec_pow:integer_to_scientific(NewTargetInt)).
```

Comment:
Is min/2 correct for the scientific representation? Just checking.

Reply:
Yes